### ISSUE TITLE
Support for CRI/Containerd on the Nodes

### ISSUE THREAD

**Issue by yvespp** [rancher/rancher#16246]
**What kind of request is this (question/bug/enhancement/feature request):** feature request

Currently [Docker is required on the Nodes](https://rancher.com/docs/rancher/v2.x/en/installation/requirements/). It would be nice if Rancher would require a CRI compatible runtime instead so we could use containerd. There are also other interesting container runtimes that support CRI: [containerd](https://github.com/kubernetes-sigs/cri-o), [cri-o](https://github.com/kubernetes-sigs/cri-o) (supported in OpenShift), [kata-containers](https://github.com/kata-containers/runtime), [frakti](https://github.com/kubernetes/frakti) etc.

As Rancher aims to support clusters brought up by other tools (GKE, AKS, EKS etc.) it would make sens to rely on CRI and not require Docker on those Nodes.
I'm not asking for full support (RKE support, logging integration etc.) but just so that Rancher can work with Nodes not using Docker.

**Side notes**

To experiment I set up [Rancher 2.1 via Helm Chart](https://rancher.com/docs/rancher/v2.x/en/installation/ha/) on Minikube with containerd and deactivated docker:
* The cluster agent worked fine
* The node agent requires access to the Docker socket to read [Docker info](https://github.com/rancher/rancher/blob/a84766bb0b04171255861223e373f781dc6e74e0/pkg/agent/node/node.go#L48) (only?) -> could be replaced by [cri status](https://github.com/kubernetes-sigs/cri-tools/blob/master/cmd/crictl/info.go#L56)
* The rancher GUI worked fine

What else explicitly requires Docker?

**Comment by superseb** [comment]
We are looking at possibilities to support other container runtimes, at this point only Docker is supported, mainly because everything regarding containers outside of k8s is done via the Docker socket. 

ref https://github.com/rancher/rke/issues/94

**Comment by wzrdtales** [comment]
@superseb The thing with containerd is, since docker uses it too, support for newer docker versions should be given now.

**Comment by egernst** [comment]
Curious if there are any updates or visibility into how this fits into roadmap @superseb, since the 2.0 version label is a bit confusing now. Thanks!

**Comment by wzrdtales** [comment]
Yah would be interesting, last time the statement was 

> at the moment we are just not seeing any demand it but as soon as it does we will most likely support.

Hope something has already started rolling there :)

**Comment by krsna1729** [comment]
2.2 is supposed to support any CRI in this video
https://youtu.be/WYPd7i15XOg?t=3869

**Comment by dal00** [comment]
> 2.2 is supposed to support any CRI as per this video
> https://youtu.be/WYPd7i15XOg?t=3869

Would be nice to get clarification on this since 2.2 is released now and I can't find any references to this in the changelogs.

**Comment by yvespp** [comment]
Relevant change: https://github.com/rancher/rancher/pull/18461

**Comment by Yajo** [comment]
So, how should be Rancher installed and configured to make use of other CRI?

**Comment by wzrdtales** [comment]
I think this is more RKE related now than rancher itself.

**Comment by krsna1729** [comment]
@egernst tried this out yesterday. Basically create your cluster with CRI of choice and import it in rancher UI. it detects the runtime. You will be able to install monitoring and other stuff from the UI

**Comment by isavcic** [comment]
@krsna1729 but the agent running on a containerd node (no docker) is looking for docker.sock and fails to start. Any workaround for this?

**Comment by egernst** [comment]
What version on rancher are you using @isavcic?

I was able to import a containerd based cluster without issue utilizing 2.2

**Comment by isavcic** [comment]
@egernst Works on v2.2.2 (after I've did a workaround for #18679), but it seems that Rancher did a "partial" import, ie. it didn't detect any Workloads in the System project.

**Comment by isavcic** [comment]
@egernst ie. I don't see the kube-system namespace, for example, nor any else which exist on the k8s being imported. Looks like some permission issue?

**Comment by isavcic** [comment]
Strike that, figured it out. Missing annotation.

**Comment by shinebayar-g** [comment]
Does Rancher support containerd as runtime?

**Comment by cjellick** [comment]
Not directly @shinebayar-g  BUT k3s (see https://k3s.io) does by default. You can deploy a k3s cluster and then import that into Rancher.

**Comment by shinebayar-g** [comment]
@cjellick I see, then **containerd** support looks like isn't on the roadmap of Rancher right?
My requirement is I have to enable [authorized cluster endpoint](https://rancher.com/docs/rancher/v2.x/en/cluster-provisioning/rke-clusters/options/#authorized-cluster-endpoint) so if I import k3s cluster I wouldn't have this feature right?

**Comment by finkr** [comment]
There are many other CRI-O candidates. gVisor was discussed on [Rancher blog in 2018](https://rancher.com/blog/2018/2018-05-24-what-is-gvisor/)

**Comment by egernst** [comment]
(Late) Note: kata integrates well w both containerd and crio.

**Comment by mitchellmaler** [comment]
@cjellick Now that development has started on the Alpha RKE v2, will direct CRI support be part of that?

**Comment by cjellick** [comment]
Similar to k3s, RKE2 will support an embedded containerd and the ability to swap that out with a CRI of your choice. If you do swap out, RKE2's "support" will stop at the CRI interface. You the operator will have to fully manage the CRI deployment and operations yourself.

**Comment by wzrdtales** [comment]
> You the operator will have to fully manage the CRI deployment and operations yourself.

What would that mean exactly? I assume you mean just the minor task of deploying and updating the CRI right?

**Comment by cjellick** [comment]
Right. Just wanted to call out that RKE2 would do any of that for you
